src/invidious/jsonfilter.cr

```diff
@@ -0,0 +1,251 @@
+
+        offset = raw_match.starts_with?(',') ? 1 : 0
+
+        opening_bracket_index = (text_index + group_name.size) + offset
+        closing_bracket_index = bracket_pairs[opening_bracket_index]
+        content_start = opening_bracket_index + 1
+
+        content = fields_text[content_start...closing_bracket_index]
+
+        if content.empty?
+          raise FieldsParser::ParseError.new "Empty nest group at position #{content_start}"
+        else
+          content = remove_nest_groups(content)
+        end
+
+        while nest_stack.size > 0 && closing_bracket_index > nest_stack[nest_stack.size - 1][:closing_bracket_index]
+          if nest_stack.size
+            nest_stack.pop
+          end
+        end
+
+        group_name.split('/').each do |group_name|
+          nest_stack.push({
+            group_name:            group_name,
+            closing_bracket_index: closing_bracket_index,
+          })
+        end
+
+        if !content.empty?
+          properties = content.split(',')
+
+          properties.each do |prop|
+            nest_list = nest_stack.map { |nest_prop| nest_prop[:group_name] }
+
+            if !prop.empty?
+              if prop.includes?('/')
+                parse_single_nests(prop) { |list| nest_list += list }
+              else
+                nest_list.push prop
+              end
+            else
+              raise FieldsParser::ParseError.new "Empty key in nest list: #{nest_list << prop}"
+            end
+
+            yield nest_list
+          end
+        end
+      end
+    end
+
+    def self.remove_nest_groups(text : String) : String
+      content_bracket_pairs = get_bracket_pairs(text, false)
+
+      content_bracket_pairs.each_key.to_a.reverse.each do |opening_bracket|
+        closing_bracket = content_bracket_pairs[opening_bracket]
+        last_comma = text.rindex(',', opening_bracket) || 0
+
+        text = text[0...last_comma] + text[closing_bracket + 1...text.size]
+      end
+
+      return text.starts_with?(',') ? text[1...text.size] : text
+    end
+
+    def self.get_bracket_pairs(text : String, recursive = true) : BracketIndex
+      istart = [] of Int64
+      d = BracketIndex.new
+
+      text.each_char_with_index do |char, index|
+        if char == '('
+          istart.push(index.to_i64)
+        end
+
+        if char == ')'
+          begin
+            opening = istart.pop
+            if recursive || (!recursive && istart.size == 0)
+              d[opening] = index.to_i64
+            end
+          rescue
+            raise FieldsParser::ParseError.new "No matching opening parenthesis at: #{index}"
+          end
+        end
+      end
+
+      if istart.size != 0
+        idx = istart.pop
+        raise FieldsParser::ParseError.new "No matching closing parenthesis at: #{idx}"
+      end
+
+      return d
+    end
+  end
+
+  class FieldsGrouper
+    alias JSONSkeleton = Hash(String, JSONSkeleton)
+    alias SkeletonValue = Hash(String, JSONSkeleton)
```

Comment:
This can be simplified to:
```crystal
alias SkeletonValue = String | Hash(String, SkeletonValue)
```

Reply:
I'm still a bit confused with this change.

The `self.create_json_skeleton` function is supposed to output a `Hash` only containing the properties that were requested. Should look something like this:
```crystal
fields_text = "fruits(name,color(test1,test2))"
puts JSONFilter::FieldsGrouper.create_json_skeleton(fields_text)
# outputs
{"fruits" => {"name" => {}, "color" => {"test1" => {}, "test2" => {}}}}
```
Unless I'm misunderstanding, I don't think `String | Hash(String, SkeletonValue)` is quite right.

Reply:
As mentioned below we instead create `root_hash` as `Hash(String, SkeletonValue).new`, rather than `SkeletonValue.new` so the return value will always be a hash.

Reply:
The `String` part of your suggestion is what confused me since `SkeletonValue` should always be a `Hash`.

I solved it by just using:
```crystal
alias SkeletonValue = Hash(String, SkeletonValue)
```
and using your other suggestion instead of `SkeletonValue.new`
```crystal
{} of String => SkeletonValue
```
Thanks for the help! :slightly_smiling_face: